bin/varnishd/http2/cache_http2_proto.c

```diff
@@ -382,7 +382,7 @@ h2_rx_priority(struct worker *wrk, struct h2_sess *h2, struct h2_req *r2)
 
 	(void)wrk;
 	ASSERT_RXTHR(h2);
-	xxxassert(r2->stream & 1);
+	(void)r2;
```

Comment:
Maybe something like this instead of not using `wrk` and `r2`?

```c
CHECK_OBJ_NOTNULL(wrk, WORKER_MAGIC);
ASSERT_RXTHR(h2);
CHECK_OBJ_ORNULL(r2, H2_REQ_MAGIC);
if (r2 != NULL)
        AN(r2->stream & 1);
```

Reply:
I only touched that line because the changes may make r2 null there and since a [odd stream number check](https://github.com/varnishcache/varnish-cache/blob/6fdc89c1fdd12bb83fadd8412faa1d7703761e36/bin/varnishd/http2/cache_http2_proto.c#L911) is already in place in h2_procframe I removed it. It looks useless to me, and a bit wrong: ok, Varnish doesn't support push promise streams, but PRIORITY frames aren't incompatible with these. Also, why perform a check like this only in h2_rx_priority and not in any other h2_rx_*?
I digged into commit history: that assert [got added in the very first commit](https://github.com/varnishcache/varnish-cache/blob/66d2682647b8fc00c32f69bb4b1856d7d4dcc30c/bin/varnishd/http2/cache_http2_proto.c#L333) of HTTP2, at the time there wasn't the check in h2_procframe (in fact, there was no h2_procframe, only h2_rxframe) but a similar assert was [inside h2_rx_headers](https://github.com/varnishcache/varnish-cache/blob/66d2682647b8fc00c32f69bb4b1856d7d4dcc30c/bin/varnishd/http2/cache_http2_proto.c#L395). Then, [a few months later](https://github.com/varnishcache/varnish-cache/commit/77cab3676155ac3f45db4176795f5621e12be8e6) the assert in h2_rx_headers got removed and [the check we have today appeared in h2_rxframe](https://github.com/varnishcache/varnish-cache/blob/77cab3676155ac3f45db4176795f5621e12be8e6/bin/varnishd/http2/cache_http2_proto.c#L524-L531), it got essentially untouched since then.
I think the assert in h2_rx_priority should have been removed in that same commit, along with the one in h2_rx_headers.

If we do not want to leave h2_rx_priority completely empty we could add something else: for example we could enforce [rfc7540,l,1952,1953](https://tools.ietf.org/html/rfc7540#section-6.3)
```patch
 static h2_error v_matchproto_(h2_frame_f)
 h2_rx_priority(struct worker *wrk, struct h2_sess *h2, struct h2_req *r2)
 {

        (void)wrk;
        ASSERT_RXTHR(h2);
        (void)r2;
+       if (h2->rxf_len != 5) // rfc7540,l,1952,1953
+               return (H2SE_FRAME_SIZE_ERROR);
        return (0);
 }
```

The CHECK_OBJ_* asserts look nice, however if we are going to add them I think they should go in a commit of their own. Also, for consistency they should be in every h2_rx_* function, not only here. In fact, i resorted to `(void)r2;` because h2_rx_push_promise and h2_rx_goaway (which do not use r2) do so.

Reply:
I wasn't asking this to you in particular, it just caught my eye during the review (so rather a comment for fellow reviewers=). And since we already check the stream number beforehand, then the `xxxassert` statement could really be turned into a definitive assertion.

Anyway, thanks for the thorough research, and like I said: LGTM!